### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by Nipol**
1. Solidity 0.8 automatically checks for overflows for increasing numbers and subside, removed the Counter.
2. Perspective of the EVM, the loop works on the stack, so we changed the Loop to be accessed from the memory area.
3. By default, parameters are allocated to memory, so we changed them to calldata so that importing data directly from Transaction.
4. Instead of `uint40`, use timestamp as 32bits so that it appears only as a second.
5. ~~Rather than using `call` when transferring assets, using `transfer` can use stack areas more neatly and prevent possible remote data calls.~~
6. By organizing the order of data stored in the contract, you can save the number of calls to the `sstore` and the `sload`.
7. Custom errors reduce the overall contract size by eliminating the need to store messages in the contract.
8. The process of fetching metadata has been optimized by separating metadata and description and allowing it to be called as a single `sload`.
9. Since subtracting one number from the ERC721 implementation was validated by Solidity, adding one was not overflow checked.
10.  Optimized the array associated with the allowlist. Use more storage areas, but not O(n).

### CODE REVIEW

**Comment by assafom** [comment]
> 5. Rather than using `call` when transferring assets, using `transfer` can use stack areas more neatly and prevent possible remote data calls.

I suggest keeping it as `call`. `transfer` will make it not work with smart contracts (such as multisigs), might break in the future due to gas changes, and there are other ways to protect from reentrancy. See [here](https://consensys.net/diligence/blog/2019/09/stop-using-soliditys-transfer-now/) for an article about this recommendation from Consensys Diligence.

**Comment by Nipol** [comment]
> > 5. Rather than using `call` when transferring assets, using `transfer` can use stack areas more neatly and prevent possible remote data calls.
> 
> I suggest keeping it as `call`. `transfer` will make it not work with smart contracts (such as multisigs), might break in the future due to gas changes, and there are other ways to protect from reentrancy. See [here](https://consensys.net/diligence/blog/2019/09/stop-using-soliditys-transfer-now/) for an article about this recommendation from Consensys Diligence.

Oh! I forgot. I'm going to back. Thanks

**Comment by contract-whale** [comment]
> **5. Rather than using call when transferring assets, using transfer can use stack areas more neatly and prevent possible remote data calls.**

I totally agree that "call" is suitable for Smart Contract design patterns. It is designed for more individual interactions between smart contracts and a much more recommended method to send Ether from a smart contract.

Problems with send() and transfer()
Both functions were considered the go-to solution for making payments to EOAs or other smart contracts. But since the amount of gas they forward to the called smart contract is very low, the called smart contract can easily run out of gas. This would make the payment impossible.

The problem here is that even if the receiving contract is designed carefully not to exceed the gas limit, future changes in the gas costs for some opcodes can break this design.

Therefore, the recommended way of sending payments is the function call().

**Comment by Nipol** [comment]
I'm done.

**Comment by sebayaki** [comment]
Wow.. so many fixes! Thanks a lot for your detailed review :) 
I guess I'll have to take the whole day tomorrow studying your fixes 😛

**Comment by Nipol** [comment]
11. Deploying NFT first, and deploying Factory reduces deployment size!
Seperate the NFT Mint interface from the factory, and the factory no longer has all the NFT bytecodes.

**Comment by Nipol** [comment]
12. NFT is able to receive Native Currency without any operation, so fix this.

File: contracts/DixelClubV2NFT.sol [hunk 1]
```diff
@@ -76,51 +92,52 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
...
     }
 
-    function _mintNewEdition(address to, uint24[PALETTE_SIZE] memory palette) private {
+    function _mintNewEdition(address to, uint24[PALETTE_SIZE] calldata palette) private {
         // We cannot just use balanceOf to create the new tokenId because tokens
         // can be burned (destroyed), so we need a separate counter.
-        uint256 tokenId = _tokenIdTracker.current();
+        uint256 tokenId;
+        unchecked {
+            tokenId = _tokenIdTracker++;
```
Comment (Nipol) [inline contracts/DixelClubV2NFT.sol:135]: for the security,
leave that line, even if re-enterant, the token number is increased by one more.

**Comment by sebayaki** [comment]
Will merge this into `optimization-by-nipol` branch and will check-pick each commit one-by-one while I'm studying about the optimization.

File: contracts/DixelClubV2NFT.sol [hunk 2]
```diff
@@ -25,7 +25,7 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
     uint256 private _tokenIdTracker;
     IDixelClubV2Factory private _factory;
 
-    uint40 private _initializedAt;
+    uint32 private _initializedAt;
```
Comment (sebayaki) [inline contracts/DixelClubV2NFT.sol]: uint32 supports up to Feb 2106, and that's only 84 years from now.. I guess I could live up to 300 years old, so I want this app to be running at least until I die.. 😛

anyway, I guess this doesn't save any gas here because it takes one minimum storage there (32 bytes).

File: contracts/DixelClubV2NFT.sol [hunk 3]
```diff
@@ -236,9 +236,9 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
 
     // MARK: - Update metadata
 
-    function updateMetadata(bool whitelistOnly, bool hidden, uint24 royaltyFriction, uint40 mintingBeginsFrom, uint256 mintingCost) external onlyOwner {
+    function updateMetadata(bool whitelistOnly, bool hidden, uint24 royaltyFriction, uint32 mintingBeginsFrom, uint256 mintingCost) external onlyOwner {
         require(royaltyFriction <= MAX_ROYALTY_FRACTION, "INVALID_ROYALTY_FRICTION");
-        require((_metaData.mintingBeginsFrom == mintingBeginsFrom || block.timestamp < _metaData.mintingBeginsFrom), "CANNOT_UPDATE_MITING_TIME_ONCE_STARTED");
+        require((_metaData.mintingBeginsFrom == mintingBeginsFrom || uint32(block.timestamp) < _metaData.mintingBeginsFrom), "CANNOT_UPDATE_MITING_TIME_ONCE_STARTED");
```
Comment (sebayaki) [inline contracts/DixelClubV2NFT.sol]: Do I have to down-cast `block.timestamp` here explicitly? or is it a default behavior?

File: contracts/DixelClubV2NFT.sol (hunk 2 above)
Comment (Nipol) [reply contracts/DixelClubV2NFT.sol]: Haha, this appears to be used only for whether the contract has been initialized.
Metadata has 4 bits of space left. You can use this slot to save more storage space!

File: contracts/DixelClubV2NFT.sol (hunk 3 above)
Comment (Nipol) [reply contracts/DixelClubV2NFT.sol]: This is because the `mintingBeginsFrom` field of metadata is fixed to 32bits, so it must be compared in the same type.

File: contracts/DixelClubV2NFT.sol [hunk 4]
```diff
@@ -143,50 +159,74 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
...
-        require(_metaData.whitelistOnly, "COLLECTION_IS_PUBLIC");
+    function removeWhitelist(address[] calldata list) external onlyOwner {
+        if(!_metaData.whitelistOnly) revert DixelClubV2__PublicCollection();
 
         uint256 length = list.length; // gas saving
-        for (uint256 i = 0; i < length; i++) {
+        for (uint256 i; i != length;) {
             _removeWhitelist(list[i]);
+            unchecked {
+                i++;
```
Comment (contract-whale) [inline contracts/DixelClubV2NFT.sol:205]: Change to `++i` to take advantage of pre-increment operation.

File: contracts/DixelClubV2Factory.sol [hunk 5]
```diff
@@ -47,20 +47,21 @@ contract DixelClubV2Factory is Constants, Ownable {
     function createCollection(
         string calldata name,
         string calldata symbol,
+        string calldata description,
```
Comment (sebayaki) [inline contracts/DixelClubV2Factory.sol:62]: Could you explain how separating description from MetaData struct can save SLOAD?

Comment (Nipol) [reply contracts/DixelClubV2Factory.sol:62]: The description causes overhead in the off-chain. However, metadata is often read on-chain, and if the description length exceeds 256 bits, multiple 'sload's occur to read metadata on on-chain logic.

Even if you are not worried about this, it is recommended to resize the metadata itself to be read as a single 'sload' because it is loaded in the on-chain logic.

File: contracts/DixelClubV2NFT.sol (hunk 4 above)
Comment (Nipol) [reply contracts/DixelClubV2NFT.sol:205]: Oh i missed!
You are very Good reviewer!

File: contracts/DixelClubV2Factory.sol (hunk 5 above)
Comment (Nipol) [reply contracts/DixelClubV2Factory.sol:62]: fyi evm performs sload and sstore every 256 bits!

File: contracts/DixelClubV2NFT.sol [hunk 6]
```diff
@@ -156,25 +164,30 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
 
         uint256 length = list.length; // gas saving
         for (uint256 i; i != length; ) {
-            _whitelist.push(list[i]);
+            address curr = list[i]; // gas saving
+            PosAndC storage pc = _whitelistData[curr];
+
+            if (!_whitelistData[address(0)].hitted[curr]) {
```
Comment (sebayaki) [inline contracts/DixelClubV2NFT.sol:171]: Could you explain why you used `mapping(address => bool) hitted;` here?
If you wanted to prevent adding the same wallet address twice in the `_whitelist` array, we could probably just check
 ```solidity
if (!pc.init) {
    _whitelist.push(curr);
    (pc.Position, pc.init) = (uint96(_whitelist.length) - 1, true);
}
```

File: contracts/DixelClubV2NFT.sol [hunk 7]
```diff
@@ -38,13 +38,21 @@ contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
         uint24[PALETTE_SIZE] palette; // 24bit color (16,777,216) - up to 16 colors
     }
 
+    struct PosAndC {
+        uint96 Position;
+        uint96 Count;
+        bool init;
```
Comment (sebayaki) [inline contracts/DixelClubV2NFT.sol:44]: I guess we don't need `init` separately because `count == 0` is basically "not in whitelist"

File: contracts/DixelClubV2NFT.sol (hunk 6 above)
Comment (Nipol) [reply contracts/DixelClubV2NFT.sol:171]: That's right! It's to check the duplicate address. Keeping the array short is also used to reduce read overhead.
Targeted to O(log(n))

File: contracts/DixelClubV2NFT.sol (hunk 7 above)
Comment (Nipol) [reply contracts/DixelClubV2NFT.sol:44]: What I considered when I wrote that is, This is to make sure that the user is on the AllowList and has no lefted the count.
I didn't implement `init` at first, so it's good to think of it as additional information.

File: README.md [hunk 8]
```diff
@@ -70,15 +70,15 @@ npx hardhat verify --network bscmain {contract address} "parameter 1" "parameter
 ·······················|························|·············|·············|··············|···············|··············
 |  DixelClubV2Factory  ·  updateImplementation  ·          -  ·          -  ·       28949  ·            3  ·          -  │
 ·······················|························|·············|·············|··············|···············|··············
-|  DixelClubV2NFT      ·  addWhitelist          ·      56672  ·     218145  ·      108830  ·           29  ·          -  │
+|  DixelClubV2NFT      ·  addWhitelist          ·      39044  ·     189236  ·      167735  ·           29  ·          -  │
```
Comment (sebayaki) [inline README.md]: Oh it is actually much cheaper than using `EnumerableMapping` I tried in the first place, but it seems it's still ~60% more expensive than a simple array implementation.

I'm thinking of a hacky solution @assafom  suggested here: #8 
What do you think about that one?

Comment (Nipol) [reply README.md]: The maximum minimum gas requirement has been reduced, and the average is high.

Comment (Nipol) [reply README.md]: I think we need to check the test case that manages more than 100 lists. The number of lists is too small to check the edge case at this time.
It seems difficult to check if this is a more expensive logic now.

Comment (sebayaki) [reply README.md]: True. I'll add some test cases for a big list first and let's see.

Comment (sebayaki) [reply README.md]: Added some benchmark test:
https://github.com/Steemhunt/dixel-v2-contract/commit/6495dd6b59ce92cfce76a26d222bc1a3beccfb5a

Threads for whitelist optimization:
https://github.com/Steemhunt/dixel-v2-contract/issues/2#issuecomment-1146777315

**Comment by Nipol** [comment]
Ethereum, yoonsung.eth `0x4Fe992E566F8a28248acC4cB401b7FfD7dF959B0`
😉
